File: 2025-10-October-GFG-POTD/October 07 - Bottom View of Binary Tree/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

class Solution {
public:
    vector<int> bottomView(Node* root) {
        vector<int> result;
        if (!root) return result;

        map<int, int> mp; // horizontal distance -> node value
        queue<pair<Node*, int>> q; // (node, hd)
        q.push({root, 0});

        while (!q.empty()) {
            auto [node, hd] = q.front();
            q.pop();

            // Update map (latest node at hd)
            mp[hd] = node->data;

            if (node->left) q.push({node->left, hd - 1});
            if (node->right) q.push({node->right, hd + 1});
        }

        for (auto& it : mp)
            result.push_back(it.second);

        return result;
    }
};

File: 2025-10-October-GFG-POTD/October 07 - Bottom View of Binary Tree/Solution.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    def bottomView(self, root):
        result = []
        if not root:
            return result

        hd_map = {}  # horizontal distance -> node value
        queue = deque([(root, 0)])  # (node, hd)

        while queue:
            node, hd = queue.popleft()

            # Update the map (latest node at this hd)
            hd_map[hd] = node.data

            if node.left:
                queue.append((node.left, hd - 1))
            if node.right:
                queue.append((node.right, hd + 1))

        # Return values sorted by horizontal distance
        for hd in sorted(hd_map.keys()):
            result.append(hd_map[hd])

        return result
